airflow/models.py

```diff
@@ -3455,14 +3366,176 @@ def find(dag_id, run_id=None, state=None, external_trigger=None, session=None):
         qry = session.query(DR).filter(DR.dag_id == dag_id)
         if run_id:
             qry = qry.filter(DR.run_id == run_id)
+        if execution_date:
+            qry = qry.filter(DR.execution_date == execution_date)
         if state:
             qry = qry.filter(DR.state == state)
         if external_trigger:
             qry = qry.filter(DR.external_trigger == external_trigger)
-        dr = qry.all()
+        dr = qry.order_by(DR.execution_date).all()
 
         return dr
 
+    @provide_session
+    def get_task_instances(self, state=None, session=None):
+        """
+        Returns the task instances for this dag run
+        """
+        TI = TaskInstance
+        tis = session.query(TI).filter(
+            TI.dag_id == self.dag_id,
+            TI.execution_date == self.execution_date,
+        )
+        if state:
+            if isinstance(state, six.string_types):
+                tis = tis.filter(TI.state == state)
+            else:
+                # this is required to deal with NULL values
+                if None in state:
+                    tis = tis.filter(or_(
+                        TI.state.in_(state),
+                        TI.state.is_(None))
+                    )
+                else:
+                    tis = tis.filter(TI.state.in_(state))
+
+        return tis.all()
+
+    @provide_session
+    def get_task_instance(self, task_id, session=None):
+        """
+        Returns the task instance specified by task_id for this dag run
+        :param task_id: the task id
+        """
+
+        TI = TaskInstance
+        ti = session.query(TI).filter(
+            TI.dag_id == self.dag_id,
+            TI.execution_date == self.execution_date,
+            TI.task_id == task_id
+        ).one()
+
+        return ti
+
+    def get_dag(self):
+        """
+        Returns the Dag associated with this DagRun
+        :param session: database session
+        :return: DAG
+        """
+        if not self.dag:
+            raise AirflowException("{}.dag needs to be set".format(self))
+
+        return self.dag
+
+    @provide_session
+    def update_state(self, session=None):
+        """
+        Determines the overall state of the DagRun based on the state
+        of its TaskInstances.
+        :returns State:
+        """
+
+        dag = self.get_dag()
+        tis = self.get_task_instances(session=session)
+
+        logging.info("Updating state for {} considering {} task(s)".format(self, len(tis)))
+
+        for ti in tis:
+            ti.task = dag.get_task(ti.task_id)
+
+        # pre-calculate
+        # db is faster
+        # todo: this can actually get pretty slow: one task costs between 0.01-015s
+        start_dttm = datetime.now()
+        unfinished_tasks = self.get_task_instances(
+            state=State.unfinished(),
+            session=session
+        )
+        has_unfinished_tasks = True if unfinished_tasks else False
```

Comment:
It looks like this might not be needed, though maybe it makes things more clear. For example, below:

`all_deadlocked = (has_unfinished_tasks and no_dependencies_met)`
could be:
`all_deadlocked = (unfinished_tasks and no_dependencies_met)` since it will implicitly call `bool(unfinished_tasks)`

Similarly 
`elif has_unfinished_tasks and none_depends_on_past and no_dependencies_met:`
Could be
`elif unfinished_tasks and none_depends_on_past and no_dependencies_met:`